test/e2e/framework/network/utils.go

```diff
@@ -621,7 +620,10 @@ func (config *NetworkingTestConfig) setup(selector map[string]string) {
 	framework.ExpectNoError(err)
 	config.ExternalAddr = e2enode.FirstAddress(nodeList, v1.NodeExternalIP)
 
-	e2eskipper.SkipUnlessNodeCountIsAtLeast(2)
```

Comment:
> this PR proposes a more reliable way to detect available nodes for testing, that relying on a cloud config.

hi, could you explain why you think that the direct ginkgo skip is more reliable than the e2eskipper:
https://github.com/kubernetes/kubernetes/blob/cf192e690b4134808cb1f8bbf178915a949d0d9e/test/e2e/framework/skipper/skipper.go#L45-L87

Reply:
Because this way we check the number of available nodes right from the actual cluster. as oppose to e2eskipper which takes that information from the cloud config which is not universal.

Reply:
as much as i dislike the fact that we have cloud specifics in the framework, we should point out that the CloudConfig.NumNodes already receives the same value if the user has not changed it from the default of -1:
https://github.com/kubernetes/kubernetes/blob/9821d0e47e253a73fa6038e84e9a9855ce030634/test/e2e/e2e.go#L263

i'd prefer if we keep SkipUnlessNodeCountIsAtLeast(), even if the usage of the function is unclear and potentially refactor / remove the CloudCounfig completely at some point.

Reply:
CouldConfig does not work for all possible cases, hence the reason for this PR. Or do you prefer to keep inferior implementation just for emotional reasons ;) ?? So far I have not heard any `technical objection` of getting  number of the REAL source of truth, which is the actual cluster. I do not see any reason to wait for some unclear future refactoring if it can be done here and now.
Another point, how do you explain this line then: https://github.com/kubernetes/kubernetes/blob/9821d0e47e253a73fa6038e84e9a9855ce030634/test/e2e/framework/network/utils.go#L620
Why in the same code, we are getting number of nodes from the cluster not from CLoudConfig??

Reply:
> Why in the same code, we are getting number of nodes from the cluster not from CLoudConfig??

the network config code requires the list of Node objects explicitly.

yet, i do not expect the number of nodes to change during the runtime of the test suite, which means `len(framework.TestContext.CloudConfig.NumNodes == len(nodeList.Items)` at all times.

> CouldConfig does not work for all possible cases, hence the reason for this PR. 

could you clarify the cases you have in mind?

in any case your change is fine, yet what i do not like is that we are moving away from a utility that is defined for the same purpose and used in number of other locations.

let's get a second opinion here
/cc @oomichi 
PTAL

Reply:
This change adds one risk against conformance test requirements.

Basically test skipping ways are implemented with e2eskipper.Skip*** functions and conformance tests should not call these functions since the requirements.
The check is implemented as https://github.com/kubernetes/kubernetes/blob/a9c7529f990159ef327716e7c61984d5b7197115/hack/conformance/check_conformance_test_requirements.go#L37 which searches the keyword `e2eskipper.Skip.*\(` in conformance test code.
If allowing this change, the above tool will not be able to detect violations of conformance test requirements.
In addition, I don't think this makes the code so readable.
Then it is difficult to merge this for me.

Reply:
Fine I will close this PR, but then please fix (not in non determined future but asap) so not just cloud config based kubernetes multi node clusters could be used to run e2e tests.  I tested with `local` and `skeleton` and neither worked when I was specifying -num-node=2 or 5.  I use bare metal 5 node k8s cluster with 1.17.3.

Reply:
I see, thanks for your response.
Could you report the above issue on https://github.com/kubernetes/kubernetes/issues ?
It would be helpful to get more attention from many people for the issue you are facing.

Reply:
@sbezverk could you provide more details about the issue you are facing, in an issue in this repo if possible (include steps to reproduce, etc), please?

To elaborate, if we need a "skip" somewhere then it would be preferable to use e2eskipper lib (https://github.com/kubernetes/kubernetes/blob/d7e7136bbce0fa55361c0d6ebdc40357e4caf11d/test/e2e/framework/skipper/skipper.go#L65-L87) because it does more than just `ginkgo.Skip()` (as Lubomir mentined) and because if this `setup()` function was used in other tests it would erroneously include them in places where they don't belong (i.e., conformance tests), as oomichi mentioned.

The issue here is that e2e tests do some work behind the scenes to make the cloud config work. For example, some tests begin by doing something like
```go
// If NumNodes is not specified then auto-detect how many are scheduleable and not tainted
if framework.TestContext.CloudConfig.NumNodes == framework.DefaultNumNodes {
	nodes, err := e2enode.GetReadySchedulableNodes(c)
	framework.ExpectNoError(err)
	framework.TestContext.CloudConfig.NumNodes = len(nodes.Items)
}
```

The info we need to see the change that needs to be done is whether such as step is what is causing you trouble or not.
With that info we can go and figure out what part of the e2e framework to use, to refactor, or to add.
If you could tell us exactly what you are doing and the problems you are running across it would shed light into what is needed.